### PR TITLE
Fix for https://github.com/diggerhq/digger/issues/768

### LINKED ISSUE #768: AzureCI and GitLabCI don't have Panic Recovery like GithubCI and BitbucketCI
Of the four supported CI tools only Github and Bitbucket have code that recovers from a panic, for example from GitHub CI

```
	defer func() {
		if r := recover(); r != nil {
			log.Println(fmt.Sprintf("stacktrace from panic: \n" + string(debug.Stack())))
			err := usage.SendLogRecord(githubActor, fmt.Sprintf("Panic occurred. %s", r))
			if err != nil {
				log.Printf("Failed to send log record. %s\n", err)
			}
			os.Exit(1)
		}
	}()
```
It would be great if this could be written in a manner that makes it not only support all existing CIs but also as new CIs are added this change could cover those as well.

I am going to take a stab at solving this issue myself and submitting the fix as a PR.

### PR DESCRIPTION
**Pull request by ajaxian79**
Moved and refactored panic recovery into the main func and removed from individual ci handlers for a consolidated flow.

Closes #768

### CODE REVIEW

**Comment by motatoes** [comment]
tests were not triggered for some reason so I will be closing and re-opening to try and trigger them

File: cmd/digger/main.go [hunk 1]
```diff
@@ -853,19 +819,39 @@ func main() {
...
 	switch ci {
 	case digger.GitHub:
+		logLeader = os.Getenv("GITHUB_ACTOR")
 		gitHubCI(lock, policyChecker, backendApi, reportStrategy)
 	case digger.GitLab:
+		logLeader = os.Getenv("CI_PROJECT_NAME")
 		gitLabCI(lock, policyChecker, backendApi, reportStrategy)
 	case digger.Azure:
+		azureContext := os.Getenv("AZURE_CONTEXT")
+		parsedAzureContext, _ := azure.GetAzureReposContext(azureContext)
```
Comment (motatoes) [inline cmd/digger/main.go:838]: My only concern is that in this case the work for parsing payload will be done twice

It is a minor task but we should probably convert those parsers into singletons in the future

Should be OK to merge for now but can we make a note on top of this line with a comment which says something like "TODO: refactor this function to make it not perform double work after calling the `acureCI` function"

**Review by motatoes — CHANGES_REQUESTED** [review]
Could we just add a comment on the azure section for refactoring in the future? And then its good to go 👍

**Comment by ajaxian79** [comment]
@motatoes I will definitely add a comment when I shoot a follow up video this weekend, thanks!  I may even take a stab at refactoring

**Review by motatoes — APPROVED** [review]

**Comment by motatoes** [comment]
❤️  thanks for your contribution @ajaxian79

**Comment by ajaxian79** [comment]
> ❤️  thanks for your contribution @ajaxian79 

It was my genuine pleasure and I look forward to contributing more in the future!

**Pull request merged**